include/xtl/xthreadpool.hpp

```diff
@@ -0,0 +1,286 @@
+                        desired_n_threads : 
+                        desired_n_threads == Nice ? 
+                            int(std::thread::hardware_concurrency()) / 2 : 
+                            int(std::thread::hardware_concurrency());
+        #endif
+    }
+
+    int numThreads_;
+};
+
+/********************************************************/
+/*                                                      */
+/*                      xthreadpool                      */
+/*                                                      */
+/********************************************************/
+
+    /**\brief Thread pool class to manage a set of parallel workers.
+
+        <b>\#include</b> \<xtl/xthreadpool.hpp\><br>
+        Namespace: xtl
+    */
+class xthreadpool
+{
+  public:
+
+    /** Create a thread pool from xparallel_options. The constructor just launches
+        the desired number of workers. If the number of threads is zero,
+        no workers are started, and all tasks will be executed in synchronously
+        in the present thread.
+     */
+    xthreadpool(const xparallel_options & options = xparallel_options())
+    :   options_(options),
+        stop(false)        
+    {
+        init();
+    }
+
+
+
+    /** Create a thread pool with n threads. The constructor just launches
+        the desired number of workers. If \arg n is <tt>xparallel_options::Auto</tt>,
+        the number of threads is determined by <tt>std::thread::hardware_concurrency()</tt>.
+        <tt>xparallel_options::Nice</tt> will create half as many threads.
+        If <tt>n = 0</tt>, no workers are started, and all tasks will be executed
+        synchronously in the present thread. If the preprocessor flag
+        <tt>xtl_SINGLE_THREADED</tt> is defined, the number of threads is always set
+        to zero (i.e. synchronous execution), regardless of the value of \arg n. This
+        is useful for debugging.
+     */
+    xthreadpool(const int n)
+    :   options_(n),
+        stop(false)
+    {
+        init();
+    }
+
+
+    /**
+     * The destructor joins all threads.
+     */
+    ~xthreadpool();
+
+    /**
+     * Enqueue a task that will be executed by the thread pool.
+     * The task result can be obtained using the get() function of the returned future.
+     * If the task throws an exception, it will be raised on the call to get().
+     */
+    template<class F>
+    auto enqueue(F&& f) -> std::future<decltype(f(0))>;
+
+
+    /**
+     * Block until all tasks are finished.
+     */
+    void wait()
+    {
+        std::unique_lock<std::mutex> lock(queue_mutex);
+        finish_condition.wait(lock, [this](){ return tasks.empty() && (busy == 0); });
+    }
+
+    /**
+     * Return the number of worker threads.
+     */
+    size_t n_worker_threads() const
+    {
+        return workers.size();
+    }
+
+    size_t n_threads() const
+    {
+        return workers.size() == 0 ? std::size_t(1) : workers.size();
```

Comment:
I probably missed the point of this since I find it confusing. This returns the same value whether I create a thread pool with a single thread or an empty thread pool. In the former case the total number of threads is 2 (main thread + the one created in the pool), in the latter case only 1.

Reply:
One of the main features of this threadpool is the thread_index / worker_index.
All lambdas get this index as an argument. The idea is simple: often one needs per-thread-data structures.
Consider the case when no extra thread was requested (we want to run in the main thread).
The thread_index / worker_index is 0 in this case (it would also be 0 with just a asingle worker).
``` c++
struct PerThreadData{
};
auto n_threads = 0;
xthreadpool pool(n_threads);
std::vector<PerThreadData> vec(pool.n_threads());

pool.enqueue([&](auto worker_index){
   // get per thread data
   auto & data = vec[worker_index];
});

```
if `pool.n_threads()`  needs to return something non-negative.
But I somehow agree that this might be too confusing. 
Do you have a better Idea how to call this function?
Or one needs to write:
``` c++
std::vector<PerThreadData> vec(pool.n_threads() == 0 ? 1 : pool.n_threads();

```